crates/auth/src/auth/permissions/types.rs

```diff
@@ -0,0 +1,599 @@
+        );
+
+        // Test application permissions
+        assert_eq!(
+            Permission::from_str("application:list[app1,app2]"),
+            Some(Permission::Application(ApplicationPermission::List(
+                ResourceScope::Specific(vec!["app1".to_string(), "app2".to_string()])
+            )))
+        );
+
+        // Test context permissions
+        assert_eq!(
+            Permission::from_str("context:execute[ctx1]<user1>method1"),
+            Some(Permission::Context(ContextPermission::Execute(
+                ResourceScope::Specific(vec!["ctx1".to_string()]),
+                UserScope::Specific("user1".to_string()),
+                Some("method1".to_string())
+            )))
+        );
+    }
+
+    #[test]
+    fn test_permission_satisfaction() {
+        // Test master permission
+        let master = Permission::Admin(AdminPermission);
+        let app_list = Permission::Application(ApplicationPermission::List(ResourceScope::Global));
+        assert!(master.satisfies(&app_list));
+
+        // Test specific application permission
+        let app_list_specific =
+            Permission::Application(ApplicationPermission::List(ResourceScope::Specific(vec![
+                "app1".to_string(),
+                "app2".to_string(),
+            ])));
+        let app_list_required =
+            Permission::Application(ApplicationPermission::List(ResourceScope::Specific(vec![
+                "app1".to_string(),
+            ])));
+        assert!(app_list_specific.satisfies(&app_list_required));
+        assert!(!app_list_required.satisfies(&app_list_specific));
+    }
+
+    #[test]
+    fn test_alias_permission_parsing() {
+        // Test create permission
+        let perm = Permission::from_str("context:alias:create").unwrap();
+        assert!(matches!(
+            perm,
+            Permission::Context(ContextPermission::Alias(AliasPermission::Create))
+        ));
+
+        // Test delete permission
+        let perm = Permission::from_str("context:alias:delete").unwrap();
+        assert!(matches!(
+            perm,
+            Permission::Context(ContextPermission::Alias(AliasPermission::Delete))
+        ));
+
+        // Test lookup permission with context
+        let perm = Permission::from_str("context:alias:lookup[ctx-123]").unwrap();
+        match perm {
+            Permission::Context(ContextPermission::Alias(AliasPermission::Lookup {
+                context_id,
+                user_id,
+            })) => {
+                assert_eq!(context_id.unwrap(), "ctx-123");
+                assert!(user_id.is_none());
+            }
+            _ => panic!("Wrong permission type"),
+        }
+
+        // Test lookup permission with context and user
+        let perm = Permission::from_str("context:alias:lookup[ctx-123,user-456]").unwrap();
+        match perm {
+            Permission::Context(ContextPermission::Alias(AliasPermission::Lookup {
```

Comment:
this will be different from, for example, `context:leave` which from the logic here I'd expect to be represented as

```
context:leave[ctx-123,ctx-456]<user-123,user-456>
```

right?

it's worth noting that this means either of the identities can be used to operate in either of the contexts